### ISSUE TITLE
模型存储问题

### ISSUE THREAD

**Issue by DTaoo** [PaddlePaddle/Paddle#25266]
背景：在一个pretrained backbone network上面添加一个新的fc，只优化FC层。
问题：将pretrained network 最后输出设置为stop_gradient=True, 只优化最后的fc，打印对比发现，backbone weight 并未改变（符合预期），但是将此backbone用save_inference_model函数存储后，在数据集上测试结果发生异常（acc:0.779)。
如果不进行网络优化，直接存储backbone，在数据集上测试结果为acc 0.814。

0.779 vs. 0.814的主要区别在于下方代码是否运行：

    for pass_id in range(EPOCH_NUM):
        count = 0
        for data_train in train_loader:
            print(count)
            count += 1
            avg_loss_value = exe.run(main_program,
                                     feed=data_train,
                                     fetch_list=[train_loss])
            print(avg_loss_value)

但是打印发现error =0.000000， 表明权重未发生改变，为何会出现这样的测试差异呢？

系统环境：PaddleCould v1.7.1

from __future__ import print_function
import paddle
import paddle.fluid as fluid
import numpy
import math
import sys
from PIL import Image
from resnet import ResNet34
import os
import cv2
import sys
import numpy as np
import six
import time
from arcmargin import ArcMarginLoss
from time import sleep

use_cuda = True
place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()

feed_order = ['image', 'label']

main_program = fluid.default_main_program()
star_program = fluid.default_startup_program()

exe = fluid.Executor(place)

os.system("mkdir data_center_v2")
os.system("cp -rf ./afs/datatang/aligned_set1.zip ./data_center_v2/")
os.system("unzip ./data_center_v2/aligned_set1.zip")

def vis_res34_model():
    image = fluid.data(name='image', dtype='float32', shape=[None, 3, 112, 112])
    label = fluid.data(name='label', dtype='int64', shape=[None, 1])
    loader = fluid.io.DataLoader.from_generator(feed_list=[image, label], capacity=16, iterable=True)

    model = ResNet34()
    img_fea = model.net(image)
    img_fea.stop_gradient = True

    prediction = fluid.layers.fc(input=img_fea, size=883, act='softmax')
    prediction.stop_gradient = True
    cost = fluid.layers.cross_entropy(input=prediction, label=label)
    loss = fluid.layers.mean(cost)
    
    #img_fea.stop_gradient = True
    #var = np.array(fluid.global_scope().find_var('conv2d_9.w_0').get_tensor())+1
    #var_ = fluid.global_scope().find_var('conv2d_9.w_0').get_tensor()
    #reg_loss = reg_compute(var, var_)

    #metricloss = ArcMarginLoss(class_dim=1000)
    #cost, logit = metricloss.loss(img_fea, label)
    #loss = fluid.layers.mean(x=cost)

    return loss, loader, prediction, img_fea

def train_batch_data_reader():

    batch_size = 32
    img_data = np.load('./afs/datatang/aligned_set1_data.npy')
    img_label = np.load('./afs/datatang/aligned_set1_label.npy')

    data_dir = './aligned_set1'
    img_dir    = [os.path.join(data_dir, img) for img in img_data]

    num = int(np.floor(3140./batch_size))
    for s in range(num):
        i = s % num
        batch_image = np.zeros(shape=(batch_size, 3, 112, 112), dtype='float32')
        batch_label = np.zeros(shape=(batch_size, 1), dtype='int64')
        for j in range(batch_size):
            # print(high_imgs[i*batch_size+j])
            img = cv2.imread(img_dir[i*batch_size+j])
            #print(img_dir[i*batch_size+j])
            #print(img.shape)
            img = cv2.resize(img, (112, 112))

            mean = np.mean(img, axis=(0, 1))
            std = np.std(img, axis=(0, 1))
            img = img.astype(np.float32)
            for k in range(len(std)):
                if std[k] == 0:
                    std[k] = 1
            for c in range(3):
                img[:, :, c] = (img[:, :, c] - mean[c]) / std[c]

            im = img.transpose((2, 0, 1))

            batch_image[j, :, :, :] = im
            batch_label[j,:] = img_label[i*batch_size + j]

        yield batch_image, batch_label

def optimizer_program():
    return fluid.optimizer.Momentum(learning_rate=0.01, momentum=0.9, parameter_list=['fc_2.w_0', 'fc_2.b_0'])

train_loss, train_loader, prediction, img_fea = vis_res34_model()
optimizer = optimizer_program()
optimizer.minimize(train_loss)

train_loader.set_batch_generator(train_batch_data_reader, places=place)
EPOCH_NUM = 5

params_dirname = "./output/train_fc"

#train_loader.start()

def if_exist(var):
    has_var = os.path.exists(os.path.join('./afs/vis-model/checkpoint/', var.name))
    if has_var:
        print('var: %s found' % (var.name))
    return has_var

def train_loop():

    exe.run(star_program)
    fluid.io.load_vars(executor=exe, dirname='./afs/vis-model/checkpoint2/', main_program=main_program, predicate=if_exist)
    var_list = fluid.default_main_program().all_parameters()
    #print(var_list)
    before_tuned_dict = {}
    for var in var_list:
        if var.name not in ['fc_2.w_0', 'fc_2.b_0']:
            before_tuned_dict[var.name] = fluid.global_scope().find_var(var.name).get_tensor()

    
    for pass_id in range(EPOCH_NUM):
        count = 0
        for data_train in train_loader:
            print(count)
            count += 1
            avg_loss_value = exe.run(main_program,
                                     feed=data_train,
                                     fetch_list=[train_loss])
            print(avg_loss_value)
            # save parameters
    
    after_tuned_dict = {}
    for var in var_list:
        if var.name not in ['fc_2.w_0', 'fc_2.b_0']:
            after_tuned_dict[var.name] = fluid.global_scope().find_var(var.name).get_tensor()

    error = 0.
    for var in var_list:
        if var.name not in ['fc_2.w_0', 'fc_2.b_0']:
            #print(np.sum(after_tuned_dict[var.name]))
            error += np.sum(np.abs(np.array(after_tuned_dict[var.name])-np.array(before_tuned_dict[var.name])))

    print('%.6f' % error)

    test_program = main_program.clone(for_test=True)
    fluid.io.save_inference_model(params_dirname, ["image"], [img_fea], exe)#, main_program=test_program)
    os.system("rm -rf ./afs/datatang_model/hudi/train_fc")
    os.system("cp -r ./output/train_fc ./afs/datatang_model/hudi/")

train_loop()

**Comment by zhiqiu** [comment]
同步一下结论。

经线下沟通发现是，batch_norm的参数会变化。
![image](https://user-images.githubusercontent.com/6888866/86112464-9570a800-bafa-11ea-8c0b-9bb41452d6bb.png)

原因是，backbone中使用了batch_norm，这样即使对backbone设置stop_gradient=True，batch_norm的参数mean和varience会还是在前向过程中更新。
通过设置`use_global_stats=True`可以禁止训练阶段更新这两个参数。
